pkg/collector/corechecks/snmp/internal/profile/rc_provider.go

```diff
@@ -0,0 +1,110 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+package profile
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/DataDog/datadog-agent/comp/remote-config/rcclient"
+	"github.com/DataDog/datadog-agent/pkg/networkdevice/profile/profiledefinition"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"maps"
+	"slices"
+	"sync"
+	"time"
+)
+
+var rcSingleton *UpdatableProvider
+var rcOnce sync.Once
+var rcError error
+
+// NewRCProvider returns a profile provider that subscribes to remote
+// configuration and receives profile updates from the backend. Multiple calls
+// will return the same singleton object.
+func NewRCProvider(client rcclient.Component) (Provider, error) {
+	rcOnce.Do(func() {
+		rcSingleton, rcError = buildAndSubscribeRCProvider(client)
+	})
+	return rcSingleton, rcError
+}
+
+// buildAndSubscribeRCProvider builds a new UpdatableProvider and subscribes to
+// RC to receive profile updates.
+func buildAndSubscribeRCProvider(rcClient rcclient.Component) (*UpdatableProvider, error) {
+	// Load OOTB profiles from YAML
+	defaultProfiles := getYamlDefaultProfiles()
+	if defaultProfiles == nil {
+		return nil, fmt.Errorf("could not find OOTB profiles")
+	}
+	userProfiles := make(ProfileConfigMap)
+
+	provider := &UpdatableProvider{}
+	provider.Update(userProfiles, defaultProfiles, time.Now())
+
+	// Subscribe to the RC client
+	log.Debug("Subscribing to remote config profiles")
+	rcClient.Subscribe(state.ProductNDMDeviceProfilesCustom, makeOnUpdate(provider))
```

Comment:
There will generally be multiple check instances, with new ones being created when devices are discovered; the subscription is done through a singleton instance of UpdatableProvider to ensure we only subscribe once. We could track the number of current check instances so that we can unsubscribe if it ever hits zero, but in practice this should be extremely rare - if an agent finds one device to monitor then it will almost certainly find multiple devices, and the odds of all of them going offline are low enough that I don't think we need to worry about the unnecessary extra subscription in that case.